stan/math/prim/mat/functor/mpi_parallel_call.hpp

```diff
@@ -0,0 +1,488 @@
+          throw std::domain_error("MPI error on first evaluation.");
+        } else {
+          // and ensure on the workers that they return into their
+          // listening state
+          return result_t();
+        }
+      }
+
+      // execution was ok everywhere such that we can now distribute the
+      // outputs per job as recorded on the root on the first execution
+      boost::mpi::broadcast(world_, local_outputs_per_job, 0);
+      num_outputs_per_job_ = local_outputs_per_job;
+
+      if (!cache_f_out::is_valid()) {
+        // ok, now all is fine everywhere, collect sizes and cache info
+        std::vector<int> world_f_out(num_jobs, 0);
+        boost::mpi::gatherv(world_, local_f_out.data(), num_local_jobs,
+                            world_f_out.data(), job_chunks, 0);
+        // on the root we now have all sizes from all childs. Copy
+        // over the local sizes to the world vector on each local
+        // node in order to cache this information locally.
+        std::copy(local_f_out.begin(), local_f_out.end(),
+                  world_f_out.begin() + start_job);
+        cache_f_out::store(world_f_out);
+      }
+    }
+
+    typename cache_f_out::cache_t& world_f_out = cache_f_out::data();
+
+    // check that cached sizes are the same as just collected from
+    // this evaluation
+    for (int i = 0; i < num_local_jobs; ++i) {
+      if (world_f_out[start_job + i] != local_f_out[i]) {
+        local_ok = 0;
+        break;
+      }
+    }
+
+    const std::size_t size_world_f_out = sum(world_f_out);
+    matrix_d world_result(num_outputs_per_job_, size_world_f_out);
+
+    std::vector<int> chunks_result(world_size_, 0);
+    for (std::size_t i = 0, ij = 0; i != world_size_; ++i)
+      for (int j = 0; j != job_chunks[i]; ++j, ++ij)
+        chunks_result[i] += world_f_out[ij] * num_outputs_per_job_;
+
+    // collect results on root
+    boost::mpi::gatherv(world_, local_output.data(), chunks_result[rank_],
+                        world_result.data(), chunks_result, 0);
+
+    // let root know if all went fine everywhere
+    int cluster_status = 0;
+    boost::mpi::reduce(world_, local_ok, cluster_status, std::plus<int>(), 0);
+    bool all_ok = cluster_status == static_cast<int>(world_size_);
+
+    // on the workers all is done now.
+    if (rank_ != 0)
+      return result_t();
+
+    // in case something went wrong we throw on the root
+    if (cluster_status != static_cast<int>(world_size_))
+      throw std::domain_error("Error during MPI evaluation.");
+
+    return combine_(world_result, world_f_out);
+  }
+
+ private:
+  template <typename T_cache>
+  void scatter_array_2d_cached(typename T_cache::cache_t& data) {
+    // distribute data only if not in cache yet
+    if (T_cache::is_valid()) {
+      return;
+    }
+
+    // number of elements of each data item must be transferred to
+    // the workers
+    std::vector<int> data_dims = dims(data);
+    data_dims.resize(2);
```

Comment:
I think this can just be

```
std::vector<int> data_dims(data.begin(), data.begin() + 2);
```
I think you could also just bypass the vector storage by declaring
```
int* data_dims = { data[0], data[1] };
```
But I'd really be inclined just to name the entries because they get pulled out again as `data_dims[0]` and `data_dims[1]` later and by naming them, you auto-doc them.

Reply:
Uhmm... `dims` is a stan::math function which tells me the size of the data shapes (and not the first two entries). So I guess I should leave this as is; note that different things happen on the root (sending) and on the workers (receiving).